### PR TITLE
iop-order: Move color equal before color balance rgb.

### PR DESCRIPTION
**Pull request by TurboGit**

### CODE REVIEW

**Comment by TurboGit** [comment]
@s7habo : Can you test having Color Equal before Color Balance RGB? No need to build this PR you can just change the order on GUI. If ok on your side we may want to merge this. TIA.

**Comment by s7habo** [comment]
> Can you test having Color Equal before Color Balance RGB?

Yes, that is a possibility and that's what I did. The problem is that when the saturation is low, brightness regulation of color equalizer is very weak.

In this example there is _hardly any difference_ in brightness compared to the original although _all sliders in color equalizer are at maximum_. (left original/right brightness with color equalizer):

grafik

Completely different, the color equalizer _after saturation with color balance_:

First, with the saturation of the color balance module, which is now above color equalizer, hardly anything changes in terms of brightness:

grafik

You get completely different results if you swap the positions in the pixel pipe:

grafik

In other words, the brightness function of the color equalizer is not very useful if the necessary saturation is missing.

**Comment by s7habo** [comment]
>  If ok on your side we may want to merge this. TIA.

Sorry, I forgot to answer. I also think it makes sense that the color equalizer is under the color balance module in the pixel pipe.

**Comment by TurboGit** [comment]
@jenshannoschwalm : I have tested this and seems to work better by giving a boost to the brightness:

```
diff --git a/src/iop/colorequal.c b/src/iop/colorequal.c
index 054d32f404..089611ba90 100644
--- a/src/iop/colorequal.c
+++ b/src/iop/colorequal.c
@@ -795,7 +795,7 @@ void process(struct dt_iop_module_t *self,
       const float sat = pix_out[1];
       corrections_out[0] = lookup_gamut(d->LUT_hue, hue);
       corrections_out[1] = lookup_gamut(d->LUT_saturation, hue);
-      corrections_out[2] = 16.f * sat * (lookup_gamut(d->LUT_brightness, hue) - 1.f);
+      corrections_out[2] = 64.f * sat * (lookup_gamut(d->LUT_brightness, hue) - 1.f);
     }
     else
     {
@@ -834,7 +834,7 @@ void process(struct dt_iop_module_t *self,
     /* Define a weighing function for brightness correction based on saturation
        with an inflection point at 0.15 reaching ~80% of correction at 0.2
     */
-    const float weight = 1.0f / (1.0f + expf(-(30.0f*(pix_out[1] - 0.15f))));
+    const float weight = 1.0f / (1.0f + expf(-(50.0f*(pix_out[1] - 0.10f))));
     pix_out[1] *= corrections_out[1]; // the brightness and saturation are gains
     pix_out[2] *= 1.0f + (weight * corrections_out[2]);
 

```

But for saturated colors it becomes very very sensitive, so a very little slider change do a big change of brightness.

One option would be to have a boost slider on the brightness tab to control the boost depending on saturation of the picture and so giving full control to users. This would replace the hard coded 64 above.

Or maybe a better option you can come with :)

**Comment by jenshannoschwalm** [comment]
> Or maybe a better option you can come with :)

Working on it ...

**Comment by TurboGit** [comment]
Ok, I saw that you were working on it, but I wanted to play a bit too :) I have implemented something based on sat maximum to get an auto boost factor based on overall saturation of the picture.

```
diff --git a/src/iop/colorequal.c b/src/iop/colorequal.c
index 054d32f404..791cc593a1 100644
--- a/src/iop/colorequal.c
+++ b/src/iop/colorequal.c
@@ -769,10 +769,12 @@ void process(struct dt_iop_module_t *self,
 
   // STEP 3 : carry-on with conversion from LUV to HSB
 
+  float sat_max = 0.1f;
+
 #ifdef _OPENMP
 #pragma omp parallel for simd default(none) \
-  dt_omp_firstprivate(ch, npixels, in, out, UV, L, corrections, input_matrix, d, white)  \
-  schedule(simd:static) aligned(in, out, UV, L, corrections, input_matrix : 64)
+  dt_omp_firstprivate(ch, npixels, in, out, UV, L, corrections, input_matrix, d, white) \
+  schedule(simd:static) aligned(in, out, UV, L, corrections, input_matrix : 64) reduction(max : sat_max)
 #endif
   for(size_t k = 0; k < npixels; k++)
   {
@@ -795,7 +797,8 @@ void process(struct dt_iop_module_t *self,
       const float sat = pix_out[1];
       corrections_out[0] = lookup_gamut(d->LUT_hue, hue);
       corrections_out[1] = lookup_gamut(d->LUT_saturation, hue);
-      corrections_out[2] = 16.f * sat * (lookup_gamut(d->LUT_brightness, hue) - 1.f);
+      corrections_out[2] = lookup_gamut(d->LUT_brightness, hue);
+      sat_max = MAX(sat_max, sat);
     }
     else
     {
@@ -818,7 +821,7 @@ void process(struct dt_iop_module_t *self,
   // STEP 3: apply the corrections and convert back to RGB
 #ifdef _OPENMP
 #pragma omp parallel for simd default(none) \
-  dt_omp_firstprivate(ch, npixels, out, corrections, output_matrix, white, d)  \
+  dt_omp_firstprivate(sat_max, ch, npixels, out, corrections, output_matrix, white, d) \
   schedule(simd:static) aligned(out, corrections, output_matrix: 64)
 #endif
   for(size_t k = 0; k < npixels; k++)
@@ -836,7 +839,7 @@ void process(struct dt_iop_module_t *self,
     */
     const float weight = 1.0f / (1.0f + expf(-(30.0f*(pix_out[1] - 0.15f))));
     pix_out[1] *= corrections_out[1]; // the brightness and saturation are gains
-    pix_out[2] *= 1.0f + (weight * corrections_out[2]);
+    pix_out[2] *= 1.0f + ((1.0f/sat_max) * weight * (corrections_out[2] - 1.0f));
 
     // Sanitize gamut
     gamut_map_HSB(pix_out, d->gamut_LUT, white);

```

Probably a bit fragile, but works on some images I have tested...
